routing/index_graph.cpp

```diff
@@ -1,10 +1,115 @@
-#include "index_graph.hpp"
+#include "routing/index_graph.hpp"
+
+#include "routing/restrictions_serialization.hpp"
 
 #include "base/assert.hpp"
 #include "base/exception.hpp"
 
+#include "std/algorithm.hpp"
 #include "std/limits.hpp"
 
+namespace
+{
+using namespace routing;
+
+bool IsRestricted(RestrictionVec const & restrictions, vector<SegmentEdge> const & edges,
+                  Segment const & u, Segment const & v, bool isOutgoing)
+{
+  uint32_t const featureIdFrom = isOutgoing ? u.GetFeatureId() : v.GetFeatureId();
+  uint32_t const featureIdTo = isOutgoing ? v.GetFeatureId() : u.GetFeatureId();
+
+  // Looking for at least one restriction of type No from |featureIdFrom| to |featureIdTo|.
+  if (binary_search(restrictions.cbegin(), restrictions.cend(),
+                    Restriction(Restriction::Type::No, {featureIdFrom, featureIdTo})))
+  {
+    if (featureIdFrom != featureIdTo)
+      return true;
+
+    // @TODO(bykoianko) According to current code if a feature id is marked as a feature with
+    // restrictricted uturn it's restricted to make a uturn on the both ends of the feature.
+    // Generally speaking it's wrong. In osm there's information about the end of the feature
+    // where the uturn is restricted. It's necessary to pass the data to mwm and to use it here.
+    // Set test LineGraph_RestrictionF1F1No for further details.
+    return u.GetSegmentIdx() == v.GetSegmentIdx() && u.IsForward() != v.IsForward();
+  }
+
+  // Taking into account that number of features starting and ending with the same feature id
+  // is relevantly small it's possible to ignore such cases.
+  if (featureIdFrom == featureIdTo)
+    return false;
+
+  // Looking for at least one restrictions of type Only from |featureIdFrom| to |featureIdTo|.
+  if (binary_search(restrictions.cbegin(), restrictions.cend(),
+                    Restriction(Restriction::Type::Only, {featureIdFrom, featureIdTo})))
+  {
+    return false;
+  }
+
+  // Looking for at least one restriction of type Only starting from |featureIdFrom|
+  // and with finish different from |featureIdTo|.
+  // Note. There is the following idea behind the binary_search below
+  // * if there is an item of type Only in |restrictions| starting from |featureIdFrom|
+  //   and finishing with |featureIdTo| the method returns false earlier.
+  // * binary_search below looks for a restriction of type Only starting from |featureIdFrom|.
+  //   If such restriction is found it means there is a restiction in |restrictions| of type Only
+  //   starting from |featureIdFrom| and finishing at feature id different from |featureIdTo|.
+  //   So it's impossible to move from |featureIdFrom| to |featureIdTo|.
+  auto const it = lower_bound(restrictions.cbegin(), restrictions.cend(),
+                              Restriction(Restriction::Type::Only, {featureIdFrom, featureIdTo}),
```

Comment:
We don't need to specify featureIdTo here, as the second element of m_featureIds is not used in the comparer.